RFC - Mailer

jadb opened the pull request:
This supersedes #6319 and as suggested by @markstory, adds some conventions borrowed from Rails' [ActionMailer](http://edgeguides.rubyonrails.org/action_mailer_basics.html).
## TODO
- [x] Move all email related classes out of `Cake\Network\Email` and into their own namespace. Agreed with @markstory and @lorenzo on `Cake\Mailer`, with aliases for BC.
- [x] Move transports into their own namespace at `Cake\Mailer\Transport` with aliases for BC.
- [x] Create a new `Mailer` class that would make working with emails very similar to controllers (actions, templates, model-aware).
- [ ] Create new bake command(s) to create mailer classes and/or templates.
- [x] Update cakephp/app's `config/bootstrap.php` and `config/app.default.php`.
## Usage

For documentation's sake and to better grasp the changes, here's a sample code of what would/should work when the work is finished.

```
bin/cake bake mailer user welcome,reset_password
```

which would generate the following:

```
- src/Mailer/UserMailer.php
- src/Template/Layout/Email/html/user.ctp
- src/Template/Layout/Email/text/user.ctp
- src/Template/Email/html/User/welcome.ctp
- src/Template/Email/html/User//reset_password.ctp
- src/Template/Email/text/User/welcome.ctp
- src/Template/Email/text/User/reset_password.ctp
```

`UserMailer` would look like this after some customizations:

``` php
<?php
namespace App\Mailer;

use ArrayObject;
use Cake\Event\Event;
use Cake\Mailer\Mailer;
use Cake\ORM\Entity;

class UserMailer extends Mailer
{
    public function welcome($user)
    {
        $this->subject = sprintf('Welcome %s', $user->name);
        $this->to = $user->email;
    }

    public function resetPassword($user)
    {
        $this->subject = sprintf('Reset password');
        $this->to = $user->email;
        $this->set(['token' => $user->token]);
    }

    public function implementedEvents()
    {
        return [
            'Model.afterSave' => 'onRegistration',
        ];
    }

   public function onRegistration(Event $event, Entity $entity, ArrayObject $options)
   {
       if ($entity->isNew()) {
            $this->send('welcome', [$entity]);
        }
    }
}
```

And then from the controller (or any class using `MailerAwareTrait`):

``` php
<?php
namespace App\Controller;

use Cake\Mailer\MailerAwareTrait;

class UsersController extends AppController
{
    use MailerAwareTrait;

    public function register()
    {
        ...
        $this->getMailer('User')->send('welcome', [$user]);
        ...
    }
}
```

or in a table using events:

``` php
<?php
namespace App\Model\Table;

use App\Mailer\UserMailer;
use Cake\ORM\Table;

class UsersTable extends Table
{
    public function initialize(array $config = [])
    {
        ...
        $this->eventManager->on(new UserMailer());
    }
}
```

Commit:
Move `Cake\Network\Email` to `Cake\Mailer`

Also moved all transports into `Cake\Mailer\Transport`.

Commit:
Initial work on `Mailer`

Xety commented:
:+1: I like it.

lorenzo commented on config/bootstrap.php:
```diff
@@ -19,6 +19,11 @@
+    class_alias('Cake\Mailer\Transport\DebugTransport', 'Cake\Network\Email\DebugTransport');
+    class_alias('Cake\Mailer\Email', 'Cake\Network\Email\Email');
+    class_alias('Cake\Mailer\Transport\MailTransport', 'Cake\Network\Email\MailTransport');
+    class_alias('Cake\Mailer\Transport\SmtpTransport', 'Cake\Network\Email\SmtpTransport');
```
these need to be outside the php version check

jadb replied:
Noted.

markstory commented on src/Mailer/Email.php:
```diff
@@ -1023,7 +1023,7 @@ protected function _constructTransport($name)
         }
 
-        $className = App::className($config['className'], 'Network/Email', 'Transport');
+        $className = App::className($config['className'], 'Mailer/Transport', 'Transport');
```
We should check the new and then old namespaces. Otherwise userland code will break.

jadb replied:
Noted. Will make necessary changes.

lorenzo replied:
I think it is fine if the break as long a we have some sort of upgrade command that can help you move the classes.

markstory replied:
But we don't have to have a break if we include 3 extra lines of code.

jadb replied:
I fixed this but it's not showing as outdated. See lines right below it.

markstory commented:
I like where this is going. Did you consider have `send('welcome', $user)` instead of using `__call`?

jadb commented:
@markstory no but I like it. I stuck to the `ActionMailer` API which only passes `$headers` to `send()`.

antograssiot commented on src/Mailer/Email.php:
```diff
@@ -262,7 +262,7 @@ class Email implements JsonSerializable, Serializable
      * The transport instance to use for sending mail.
      *
-     * @var \Cake\Network\Email\AbstractTransport
+     * @var \Cake\Mailer\Transport\AbstractTransport
```
\Cake\Mailer\AbstractTransport instead?

jadb replied:
Good catch!

antograssiot commented on src/Mailer/Email.php:
```diff
@@ -977,7 +977,7 @@ public function emailFormat($format = null)
      * @param string|AbstractTransport|null $name Either the name of a configured
      *   transport, or a transport instance.
-     * @return \Cake\Network\Email\AbstractTransport|$this
+     * @return \Cake\Mailer\Transport\AbstractTransport|$this
```
here also \Cake\Mailer\AbstractTransport

jeremyharris commented:
That is some tasty goodness right there. Tired of rewriting this functionality. Nice job!

HavokInspiration commented:
Looking at this and the other PR that was the starting point, I can already see myself implementing this in my apps. This is really clever.

Commit:
Add initial `Mailer` class + tests

Commit:
Move aliases outside of the php version check

See https://github.com/cakephp/cakephp/pull/6324\#discussion_r28206164

markstory commented on src/Mailer/Exception/MissingMailException.php:
```diff
@@ -0,0 +1,36 @@
+ * Missing Action exception - used when a controller action
+ * cannot be found, or when the controller's isAction() method returns false.
+ */
+class MissingMailException extends Exception
```
The doc comment doesn't match what the code does.

markstory commented on src/Mailer/Mailer.php:
```diff
@@ -0,0 +1,222 @@
+    {
+        if (!method_exists($this, $mail)) {
+            throw new MissingMailException([
+                'mailer' => $this->getName() . 'Mailer',
```
We can call the mailer methods actions.

markstory commented on tests/TestCase/Mailer/MailerTest.php:
```diff
@@ -0,0 +1,118 @@
+<?php
+namespace Cake\Test\TestCase\Mailer;
```
File header missing.

markstory commented on tests/TestCase/Mailer/MailerTest.php:
```diff
@@ -0,0 +1,118 @@
+
+class TestMailer extends Mailer
+{
+    public function getEmailForAssertion()
```
Why not have a public getter for the underlying email?

jadb replied:
I am not a fan of giving access to unneeded stuff but we could. Aside from tests, I don't see why one would need it using the mailer. The goal was to isolate mail related stuff for easier testing, etc.

markstory replied:
Ok, we've had problems with rotting test stubs in the past which is why I personally try to avoid them.

jadb replied:
Ok, will look into improving that if possible then.

Commit:
Support old namespace for now

Commit:
Fix namespace in docblocks

Commit:
Rename `mail` to  `action` + missing docblock

Commit:
phpcs

makallio85 commented:
Great job. Waiting for it.

ADmad commented on src/Mailer/Mailer.php:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    /**
+     * {@inheritdoc}
```
I don't think apigen is going to be able to inherit docblocks for php's api docs :)

ADmad commented on tests/TestCase/Mailer/MailerTest.php:
```diff
@@ -0,0 +1,128 @@
+use Cake\Mailer\Mailer;
+use Cake\TestSuite\TestCase;
+
+class TestMailer extends Mailer
```
Would be better to put this class in test app.

ADmad commented on src/Mailer/Mailer.php:
```diff
@@ -0,0 +1,222 @@
+     */
+    public function getName()
+    {
+        return str_replace('Mailer', '', join('', array_slice(explode('\\', get_class($this)), -1)));
```
If we add a public property `Mailer::$name` name could be set statically and this string manipulation would be required only if property is null.

ADmad commented on src/Mailer/Mailer.php:
```diff
@@ -0,0 +1,222 @@
+            $this->layout = Inflector::underscore($this->getName());
+        }
+
+        $this->_email = $email->profile((array)$this);
```
Since `send()` always does `$this->_email->profile((array)$this)` doing it constructor seems redundant.

jippi commented:
Super late question

Is `Mailer` the right namespace name?

We use `Controller`, `Model`, `Table`, `Template`,`Shell` etc, singular word for almost all directories, based on that, shouldn't it be just `Mail` ?

dereuromark commented:
I agree with @jippi Mail sounds better.

markstory commented:
So `Cake\Mail\Email` and `Cake\Mail\Mailer`?

jadb commented:
To clear things up a little. The namespace discussion was done over IRC with @markstory and @lorenzo present. Mailers work very similarly to controllers, hence also their methods being referred to as `actions` in exceptions and coming documentation. They are borrowed from Ruby's ActionMailer.

@jippi:
> We use Controller, Model, Table, Template,Shell etc, singular word for almost all directories, based on that, shouldn't it be just Mail ?

Mailer is singular.

> Also Validation vs. Validator - if it should be named Mailer

`Mailer` vs `Mailing`? :stuck_out_tongue: 

@dereuromark:

> Mail/Mail (like View/View vs Viewer) maybe sounds better.

Or `Control/Control` (like View/View) or maybe `View\Help`? :de:

dereuromark commented:
@jadb You convinced me :)

Commit:
Add missing docblocks

Commit:
Add static mailer name

Commit:
Remove redundant call to `Email::profile()`

Commit:
Move `TestMailer` to `TestApp`

jadb commented:
Bake command added, just missing the `actions`.

lorenzo commented on config/bootstrap.php:
```diff
@@ -22,6 +22,13 @@
+class_alias('Cake\Mailer\Transport\DebugTransport', 'Cake\Network\Email\DebugTransport');
+class_alias('Cake\Mailer\Email', 'Cake\Network\Email\Email');
+class_alias('Cake\Mailer\Transport\MailTransport', 'Cake\Network\Email\MailTransport');
+class_alias('Cake\Mailer\Transport\SmtpTransport', 'Cake\Network\Email\SmtpTransport');
```
I wonder if we should add this to the app template instead of just to the migration guide, it is a shame having to load all this classes just because.

markstory replied:
Does class_alias() actually load the classes?

ceeram replied:
In the example code it will autoload, a 3rd param of class_alias would be needed to set to false to turn off autoloading

jadb replied:
I personally prefer @lorenzo's suggestion of moving this stuff to the app template, however, worries me for already created apps which would need to manually update or having breaking changes to deal with. 

If we want, as @ceeram suggested, we can pass a 3rd param to control autoloading but why aren't we doing it for the `Text` class as well?

ceeram replied:
Moving this to app-template serves no purpose if we make the app-template require correct cakephp version which then does not need any aliassing.
The aliassing is needed in core for BC to applications using the current classnames.

ADmad replied:
:+1: to @ceeram's suggestion and doing same for `Text` class too.

markstory replied:
Moving the aliasing to the app template doesn't help people when they upgrade. Adding the 3rd as @ceeram suggested sounds good though.

lorenzo commented on tests/test_app/TestApp/Mailer/TestMailer.php:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
+ * @since         1.3.0
```
3.1.0

lorenzo commented on src/Mailer/Email.php:
```diff
@@ -1023,7 +1023,11 @@ protected function _constructTransport($name)
-        $className = App::className($config['className'], 'Network/Email', 'Transport');
+        $className = App::className($config['className'], 'Mailer/Transport', 'Transport');
+        if (!$className) {
+            $className = App::className($config['className'], 'Network/Email', 'Transport');
```
maybe add a trigger_error() here?

markstory commented:
Looks like only a few things left to fix up before this can be merged @jadb

jadb commented:
Yes, will wrap it up this week.

Commit:
Fix typo

Commit:
Alias moved classes but no autoloading

Commit:
Trigger deprecation warning

jadb commented:
Took care of all comments. The only thing I had planned (wishful thinking) was the bake commands but didn't get time to write those. Let me know if they are required for this to be merged or if they could be split into an issue of their own.

markstory commented:
Bake commands would be a separate pull request anyways as bake is a plugin.

jadb commented:
I agree (like cakephp/app) but was just a part of what I had listed to be done so didn't want to just drop the ball on it without flagging it I guess.

antoniovassell commented:
+1

ADmad commented:
Bake templates can be done later. This is good to merge.

markstory commented on src/Mailer/Mailer.php:
```diff
@@ -0,0 +1,249 @@
+    public function offsetGet($offset)
+    {
+        if (!$this->offsetExists($offset)) {
+            throw \Exception();
```
This exception should have a message.

jadb replied:
Will do.

markstory commented on tests/TestCase/Mailer/MailerTest.php:
```diff
@@ -0,0 +1,121 @@
+        $email = $this->getMockForEmail('attachments');
+        $email->expects($this->once())
+            ->method('attachments')
+            ->with([]);
```
It might be good to check these with non empty arguments to ensure arguments are passed correctly.

jadb replied:
Agreed. Will do.

markstory commented on tests/test_app/TestApp/Template/Email/html/html.ctp:
```diff
@@ -1,5 +1,5 @@
 <?php
-use Cake\Network\Email\Email;
+use Cake\Mailer\Email;
```
Why is this here?

jadb replied:
Hmm - I guess with find and replace in entire project for namespace, there was a reference to `Cake\Network\Email\Email`.

Commit:
Return NULL if undefined

Throwing an exception is not how it is done elsewhere
in the framework, so replaced that by returning NULL
for consistency.

Commit:
Use arguments to make sure they are being passed

Commit:
Revert "Alias moved classes but no autoloading"

This reverts commit d56d118 - and avoid recent PHP warnings.

Commit:
Merge remote-tracking branch 'origin/3.1' into 3.1-mailer

* origin/3.1: (305 commits)
  Update version number to 3.0.6
  New mock objects version causes our test suites to fail.
  Update FloatType.php
  Update DateTimeType.php
  Updated docBlock
  Fix empty query expressions for generating invalid SQL.
  Add tests for empty expression objects in association conditions.
  Add option to disable local XML file parsing.
  Port the Inflector fixes from #6635 to 3.0
  Fix typo in error template file name.
  Add test for ProgressHelper output with options.
  Add missing doc blcoks.
  Fix getOriginal() not preserving nulls.
  Fix incorrect doc blocks and PHPCS.
  Fixed @since in QueryExpressionTest
  Added __call() on QueryExpression to allow for and() and or() to be called transparently. Implements #6477
  Added default message for InvalidCsrfTokenException Updated thrown exception messages to be more descriptive of the cause
  replace DS with DIRECTORY_SEPARATOR in filesystem sub-package
  Added InvalidCsrfTokenException. Implements #6546
  Use ProgressHelper in i18n task.
  ...

markstory commented on src/Mailer/Mailer.php:
```diff
@@ -0,0 +1,249 @@
+     */
+    public function offsetExists($offset)
+    {
+        return in_array($offset, get_object_vars($this)) ||
```
Would `property_exists()` work here?

markstory commented on tests/test_app/TestApp/Mailer/TestMailer.php:
```diff
@@ -0,0 +1,32 @@
+<?php
+/**
+ * Test Suite Test App Logging stream class.
```
I would just remove this.

Commit:
Use `property_exists()` instead

Commit:
Remove unneeded doc line

bcrowe commented:
Are there still plans for a MailerAwareTrait?

jadb commented:
@bcrowe there is! Thanks for pointing it out, totally slipped through the cracks! Will try finishing that today.

Commit:
Add `MailerAwareTrait`

ADmad commented on src/Mailer/MailerAwareTrait.php:
```diff
@@ -0,0 +1,64 @@
+     * Returns a mailer instance.
+     *
+     * @param string $name Mailer's name.
+     * @param Cake\Mailer\Email|null $email Email instance.
```
Namespaces should start with `\`

ADmad commented on src/Mailer/MailerAwareTrait.php:
```diff
@@ -0,0 +1,64 @@
+
+        $className = App::className($name, 'Mailer', 'Mailer');
+
+        if (empty($className) || !class_exists($className)) {
```
`App::className()` already does a `class_exits()` check.

ADmad commented on tests/TestCase/Mailer/MailerAwareTraitTest.php:
```diff
@@ -0,0 +1,62 @@
+class MailerAwareTraitTest extends TestCase
+{
+
+
```
Extra empty line

Commit:
Fix discrepencies, thanks @ADmad

markstory commented on src/Mailer/MailerAwareTrait.php:
```diff
@@ -0,0 +1,64 @@
+            return $this->_mailers[$name];
+        }
+
+        if (is_null($email)) {
```
`$email === null` ?

markstory commented on src/Mailer/MailerAwareTrait.php:
```diff
@@ -0,0 +1,64 @@
+    {
+        if (isset($this->_mailers[$name])) {
+            return $this->_mailers[$name];
+        }
```
Is there any downside to always making new mailer instances? Is sharing state something people would expect?

jadb replied:
Both are fine by me. I can think of instances where I share the same instance but mostly in the same shell/controller so there are other ways to handle that.

markstory replied:
Thanks for defaulting to not sharing state. I think that will be more transparent to users. We can easily add state sharing later. While its harder to remove.

Commit:
phpcs fixes

Commit:
Replace `is_null()` by `===`

Commit:
Drop mailer instances stack

lorenzo commented on src/Mailer/Mailer.php:
```diff
@@ -0,0 +1,249 @@
+use Cake\Mailer\Exception\MissingActionException;
+use Cake\Utility\Inflector;
+
+abstract class Mailer implements ArrayAccess, EventListenerInterface
```
We usually add a header doc to each class explaining what it is

jadb replied:
Right, am so bad with documentation. Will do it.

markstory committed:
Add doc block header/example to Mailer.

markstory commented:
I added some doc blocks :smile:

jadb commented:
@markstory thanks!

lorenzo commented:
Great job @jadb :clap:

lorenzo commented:
So we now need docs and the bake task :)

jadb commented:
Thanks @lorenzo and to everyone else who helped reviewing/suggesting improvements.

@bcrowe started on docs. And yes, there are the bake tasks missing.
